### ISSUE TITLE
Nothing good comes out of a reason phrase

### ISSUE THREAD

**Issue by fluffy** [moq-wg/moq-transport#273]
Reasons phrases either have useful information in them, in which case it should not be in a reason phrase, it should be in a well defined field in the protocol. Or they do not have useful information in which case they should not be there. 

I think we should get rid of reason phrases - if we need to describe more information about why an error happened, we should put that data in well defined fields.

**Comment by afrind** [comment]
Individual Comment: 

I find value in reason phrases.  I agree that it's better to have specific error information in the error code, but often implementations have multiple reasons to send the same code, and adding the phrase enhances debugabiltity.  This is especially true once the error codes have frozen.  One implementation may find an error condition where none of the existing reason codes fit, so they are left with sending "UNKNOWN" or "PROTOCOL ERROR".  

HTTP removed per message reason phrases in H2.  This was good since they were always the same and redundant with the codes they had, or enough implementations assumed they were they wouldn't get proxied.  Note though that the "Proxy-Status" header has brought back an extensible mechanism for communicating additional context when something goes wrong.  I still miss the debug data in an H2 GOAWAY which was removed in H3. We log these strings and can group by them to find interesting interop bugs.

If nothing else, I think we should keep reason phrases around in moq for a while, because it's going to help interop testing. If we feel like we've nailed the error codes later on, we can yank them closer to the end.

**Comment by ianswett** [comment]
I agree that at least for now, these are useful.  When developing a new extension/optimization/etc, sometimes these can really help debugging the change.

**Comment by kixelated** [comment]
I think the reasons are useful for early interop. Nobody will instrument each error, nor do we have defined error codes for all permutations, so likely you'll have a big try/catch statement followed with `code=500 reason=error.messsage`.

**Comment by kixelated** [comment]
In QUIC, `CONNECTION_CLOSE` has a code/reason while `STOP_SENDING/RESET_STREAM` only has a code.

I like it and we should copy it; remove `reason` from `ANNOUNCE_*` and `SUBSCRIBE_*`.

**Comment by martinduke** [comment]
+1 for keeping these at least until interop has matured.

**Comment by alvestrand** [comment]
Seems like a rehash of the debate we have every time we design an error field.
My usual distillation is:
- If the application that gets the error can do something sensible based on it, it needs to be a structured field.
- If it's useful for debugging the problem, it's an unstructured text field (and that's OK).

The most common use I have for text messages is grepping the source to figure out which line produced the error. And that's essential both in product development and product maintenance. (Second most common use is to log the value of some parameter that made the error happen, which is incredibly important for reproducing the error, but where the application cannot and should not need to care.)

**Comment by fluffy** [comment]
Oh, and getting rid of reason code would save the length byte that needs to go in so many messages.

**Comment by fluffy** [comment]
Just for clarity on consensus ...  I do think the WG has rough consensus on what is in the draft today. This issue argue to talk about changing that.

**Comment by ianswett** [comment]
Discussed in London: Mo suggested adding a more detailed reason number so you knew exactly what type of PROTOCOL_VIOLATION it was.  Cullen said that saying its UTF-8 is potentially dangerous, so it being binary or ASCII might go smoothly.

**Comment by LPardue** [comment]
If I read the latest spec properly, the only 2 messages that carry a reason phrase are REQUEST_ERROR OR PUBLISH_DONE. Each define their own error space and neither define a generic PROTOCOL_VIOLATION.

Adding lots of codepoint that are similar in nature can make it harder for people to pick, meaning recipients have to second guess anyhow. But if people want to add some more, it seems easier at this stage to define some rather than add another separate field.

Bytes rather UTF-8 seems fine. Why make life more complicated than we need it?
